File: CodingInterviews/IsSymmetric.java
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return root == null ? true : recursion(root.left, root.right);
    }

    private boolean recursion(TreeNode left, TreeNode right) {
        if (left == null && right == null)
            return true;
        if (left == null || right == null || left.val != right.val)
            return false;
        return recursion(left.left, right.right) && recursion(left.right, right.left);
    }
}

File: LeetCode/Explore/binary-tree/IsSymmetric.cpp
#include "../main.h"

class Solution {
  public:
    bool isSymmetric(TreeNode *root) { return check(root, root); }

  private:
    bool check(TreeNode *t1, TreeNode *t2) {
        if (t1 == nullptr && t2 == nullptr)
            return true;
        if (t1 == nullptr || t2 == nullptr)
            return false;
        return t1->val == t2->val && check(t1->left, t2->right) &&
               check(t2->right, t1->left);
    }
};